### PR TITLE
op-supervisor: head pointers, refactor block processor

### PR DESCRIPTION
**Pull request by protolambda**
**Description**

Experimental draft of head block pointers instead of raw indices, and fixes of the x-head maintainer to use said block pointers.

Also includes an updated block-processor (the receipts fetcher), to be more asynchronous. I might split that out into a separate PR, TBD.

**Tests**

Work in progress.

### CODE REVIEW

File: op-supervisor/supervisor/backend/source/chain.go [hunk 1]
```diff
@@ -43,26 +45,26 @@ func NewChainMonitor(ctx context.Context, logger log.Logger, m Metrics, chainID
...
-	callback := newHeadUpdateProcessor(logger, unsafeProcessors, nil, nil)
+	callback := newHeadUpdateProcessor(logger, unsafeProcessors, safeProcessors, finalizedProcessors)
 	headMonitor := NewHeadMonitor(logger, epochPollInterval, cl, callback)
 
 	return &ChainMonitor{
-		log:         logger,
-		headMonitor: headMonitor,
+		log:            logger,
+		headMonitor:    headMonitor,
+		chainProcessor: unsafeBlockProcessor,
```
Comment (axelKingsley) [inline op-supervisor/supervisor/backend/source/chain.go:68]: Not sure if this is intentional, but the `unsafeBlockProcessor` is now embedded twice:
* Once as this new `chainProcessor`
* Once inside `headMonitor` <- `callback` <- `unsafeProcessors`

Looks like this new `chainProcessor` value is just so we call `Close` on it? There's already `headMonitor.Stop` so I'm not sure the purpose of this.

**Review by axelKingsley — COMMENTED** [review]
Summary of changes as I understand them:

## Head Pointers
Rather than use EntryIdx to represent the pointer into the database, there's now an application specific structure which captures the Block Number and Index (called "Logs Since")

## Downstream Effects
* UpdateCrossHeadsForChain now tracks the x-head position using this pointer notion. The pointer can initialize itself in position without searching around. 

## Safety Checkers Return Err
Rather than return `bool` for the safety check, they now return an error related to which kind of issue is encountered (validity or inclusion). Safety Checkers also needed to be updated to use the new storage mechanisms

## Head Storage hides Update Functions
Now the storage has explicit functions to update the named safety levels, per chain. it handles making the update operation

## Chain processors Refactor
Seems like Chain Processors are now using a channel based trigger instead of the previous `OnNewHead`. I'm not sure if these changes are very connected to the rest of the PR. I think it's so that it doesn't block?
Additionally, the `headpointer_processor`s that I wrote yesterday are gone, in favor of using the explicit `heads` update functions, wrapped into satisfying the HeadProcessor interface.

File: op-supervisor/supervisor/backend/db/db_test.go [hunk 2]
```diff
@@ -170,9 +171,9 @@ func setupStubbedForUpdateHeads(chainID types.ChainID) (*stubLogDB, *stubChecker
 	// the checkpoint starts somewhere behind the last known cross-safe head
 	checkpoint := entrydb.EntryIdx(15)
 	// the last known cross-safe head is at 20
-	cross := entrydb.EntryIdx(20)
+	cross := heads.HeadPointer{LastSealedBlockNum: 20}
 	// the local head (the limit of the update) is at 40
-	local := entrydb.EntryIdx(40)
+	local := heads.HeadPointer{LastSealedBlockNum: 40}
```
Comment (axelKingsley) [inline op-supervisor/supervisor/backend/db/db_test.go:189]: This change satisfied compilation errors for the tests but these tests are not meaningfully fixed yet.

File: op-supervisor/supervisor/backend/source/chain.go (hunk 1 above)
Comment (protolambda) [reply op-supervisor/supervisor/backend/source/chain.go:68]: Let's fix this in the follow-up PR

**Comment by protolambda** [comment]
Ok, dropped the non-supervisor changes, rebased on develop, cleaned up the truncated-hash -> common-hash type change conflicts, and fixed some linting issues.
Tests are still in a very rough spot due to refactor and changed assumptions, we're going to leave them commented, and address in the follow-up PR here: https://github.com/ethereum-optimism/optimism/pull/12058

File: op-supervisor/supervisor/backend/db/db_test.go [hunk 3]
```diff
@@ -213,39 +214,57 @@ func setupStubbedForUpdateHeads(chainID types.ChainID) (*stubLogDB, *stubChecker
...
+func (s *stubChecker) CrossHead(chainID types.ChainID) heads.HeadPointer {
 	return s.crossHeadForChain
 }
 
 // stubbed Check returns true for the first numSafe calls, and false thereafter
-func (s *stubChecker) Check(chain types.ChainID, blockNum uint64, logIdx uint32, logHash backendTypes.TruncatedHash) bool {
+func (s *stubChecker) check(types.ChainID, uint64, uint32, backendTypes.TruncatedHash) error {
 	if s.checkCalls >= s.numSafe {
-		return false
+		return fmt.Errorf("safety check failed")
```
[the remark this answers is not in the document]
Comment (protolambda) [reply op-supervisor/supervisor/backend/db/db_test.go:280]: it's in commented tests, to be changed soon. Ignoring this.

**Review by axelKingsley — APPROVED** [review]

**Pull request merged**